### PR TITLE
Implement caching layer for transaction 

### LINKED ISSUE #923: Implement caching layer for transaction
Don't need to worry about message yet. Just transaction table.
- tx hash (pk)
- timestamp
- gas information (use, limit, fee)
- sender
- status
- block height (fk)

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #923

### CODE REVIEW

File: chain/db/db.go [hunk 1]
```diff
@@ -45,8 +46,8 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 	db.Model(&DataSourceRevision{}).AddForeignKey(
 		"data_source_id",
 		"data_sources(id)",
-		"RESTRICT",
-		"RESTRICT",
+		"CASCADE",
```
Comment (taobun) [inline chain/db/db.go]: Please change it back to RESTRICT

File: chain/db/db.go [hunk 2]
```diff
@@ -93,6 +94,21 @@ func (b *BandDB) SetContext(ctx sdk.Context) {
 
 func (b *BandDB) HandleTransaction(tx auth.StdTx, txHash []byte, logs sdk.ABCIMessageLogs) {
 	msgs := tx.GetMsgs()
+
+	err := b.AddTransaction(
+		txHash,
+		b.ctx.BlockTime(),
+		b.ctx.BlockGasMeter().GasConsumed(),
```
Comment (taobun) [inline chain/db/db.go]: This is not gas used. Gasused get from response of DeliverTx

File: chain/db/db.go [hunk 3]
```diff
@@ -93,6 +94,21 @@ func (b *BandDB) SetContext(ctx sdk.Context) {
...
+		b.ctx.BlockGasMeter().GasConsumedToLimit(),
```
Comment (taobun) [inline chain/db/db.go]: tx.Fee.Gas

File: chain/db/db.go [hunk 4]
```diff
@@ -93,6 +94,21 @@ func (b *BandDB) SetContext(ctx sdk.Context) {
...
+		tx.Fee.Gas,
```
Comment (taobun) [inline chain/db/db.go]: tx.Fee.Amount

File: chain/db/transaction.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+		BlockHeight: blockHeight,
+	}
+}
+
+func (b *BandDB) AddTransaction(
+	txHash []byte,
+	timestamp time.Time,
+	gasUse uint64,
+	gasLimit uint64,
+	gasFee uint64,
```
Comment (taobun) [inline chain/db/transaction.go]: Receive as sdk.Coins

File: chain/db/table.go [hunk 6]
```diff
@@ -46,3 +46,14 @@ type DataSourceRevision struct {
 	BlockHeight    int64  `gorm:"not null"`
 	TxHash         []byte `gorm:"not null"`
 }
+
+type Transaction struct {
+	TxHash      []byte `gorm:"primary_key"`
+	Timestamp   int64  `gorm:"not null"`
+	GasUse      uint64 `gorm:"not null"`
+	GasLimit    uint64 `gorm:"not null"`
+	GasFee      uint64 `gorm:"not null"`
```
Comment (taobun) [inline chain/db/table.go]: String type

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/app/db_app.go [hunk 7]
```diff
@@ -129,6 +129,18 @@ func (app *dbBandApp) DeliverTx(req abci.RequestDeliverTx) (res abci.ResponseDel
 	}
 	if stdTx, ok := tx.(auth.StdTx); ok {
 		txHash := tmhash.Sum(req.Tx)
+
+		app.dbBand.AddTransaction(
+			txHash,
+			app.DeliverContext.BlockTime(),
+			uint64(res.GasUsed),
```
Comment (sorawit) [inline chain/app/db_app.go]: I don't think we need to do type casting. Just use whaever.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/table.go [hunk 8]
```diff
@@ -46,3 +46,14 @@ type DataSourceRevision struct {
 	BlockHeight    int64  `gorm:"not null"`
 	TxHash         []byte `gorm:"not null"`
 }
+
+type Transaction struct {
+	TxHash      []byte `gorm:"primary_key"`
+	Timestamp   int64  `gorm:"not null"`
+	GasUse      int64  `gorm:"not null"`
```
Comment (taobun) [inline chain/db/table.go]: GasUsed

File: chain/db/transaction.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		Sender:      sender.String(),
+		Success:     success,
+		BlockHeight: blockHeight,
+	}
+}
+
+func (b *BandDB) AddTransaction(
+	txHash []byte,
+	timestamp time.Time,
+	gasUse int64,
```
Comment (taobun) [inline chain/db/transaction.go]: ditto

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**